### PR TITLE
Feat/tx memo

### PR DESCRIPTION
**Pull request by NeverHappened**
https://p2pvalidator.atlassian.net/browse/LSC-39

See the https://github.com/lidofinance/lido-interchain-staking-contracts/pull/29 for contracts side of memo addition.

Added memo to SubmitTx message and to acknowledgement packet. I've tested it and it returns the memo back.

Test steps:
Use the ./gaia-wasm-zone/demo/readme.md to run test steps, but also need to add logging on acknowledgement to see memo returned back

### CODE REVIEW

File: proto/interchaintxs/v1/tx.proto [hunk 1]
```diff
@@ -52,7 +52,8 @@ message MsgSubmitTx {
   // capability for.
   string owner = 2;
   string connection_id = 3;
-  repeated google.protobuf.Any msgs = 4;
+  string memo = 4;
+  repeated google.protobuf.Any msgs = 5;
```
Comment (foxpy) [inline proto/interchaintxs/v1/tx.proto]: Please don't reorder protobuf message fields, they are sequentially enumerated for a good reason.

This is probably not going to cause any issues for our project (though I cannot prove it), but still.

Better safe than sorry.

**Review by foxpy — CHANGES_REQUESTED** [review]

File: proto/interchaintxs/v1/tx.proto (hunk 1 above)
Comment (NeverHappened) [reply proto/interchaintxs/v1/tx.proto]: Good point, actually It didn't work that way, and I redid it

**Comment by zavgorodnii** [comment]
only one consideration — maybe we shouldn't make the memo a required command line argument?.. i mean, not everyone needs a memo

File: x/interchaintxs/client/cli/tx.go [hunk 2]
```diff
@@ -81,6 +81,7 @@ func SubmitTxCmd() *cobra.Command {
 			connectionID := args[0]
 			owner := args[1]
 			pathToMsgs := args[2]
+			memo := args[3]
```
Comment (zavgorodnii) [inline x/interchaintxs/client/cli/tx.go]: I don't get it. You set MinimumNArgs(3), and here you always take the 4-th element of an array that might not have 4 elements.

Comment (ratik) [inline x/interchaintxs/client/cli/tx.go]: btw, should we check memo length here or somewhere else?

Comment (NeverHappened) [reply x/interchaintxs/client/cli/tx.go]: probably should in the message server handler. I also want to find where the real limit on memo is set in code

Comment (ratik) [reply x/interchaintxs/client/cli/tx.go]: https://github.com/cosmos/cosmos-sdk/blob/6f070623741fe0d6851d79ada41e6e2b1c67e236/x/auth/types/params.go#L13

**Review by zavgorodnii — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by ratik — APPROVED** [review]

File: x/interchaintxs/client/cli/tx.go [hunk 3]
```diff
@@ -67,10 +67,10 @@ func RegisterInterchainAccountCmd() *cobra.Command {
 
 func SubmitTxCmd() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "submit-tx [connection-id] [owner] [path/to/sdk_msgs.json]",
+		Use:     "submit-tx [connection-id] [owner] [path/to/sdk_msgs.json] [memo]",
```
Comment (foxpy) [inline x/interchaintxs/client/cli/tx.go]: Hmmm, this help message implies (people usually use square brackets to mark optional arguments) that connection-id, owner, path and memo are all optional arguments, but they are not. I am providing my own suggestion but I am not really sure about it neither.

```suggestion
		Use:     "submit-tx connection-id owner path/to/sdk_msgs.json [memo]",
```

**Review by foxpy — CHANGES_REQUESTED** [review]

File: x/interchaintxs/client/cli/tx.go (hunk 3 above)
Comment (mmikev) [reply x/interchaintxs/client/cli/tx.go]: It's a common practice in Cosmos-SDK to wrap arguments in square brackets. So I'd stick with current version

Comment (mmikev) [reply x/interchaintxs/client/cli/tx.go]: But I'd personally pass memo via flag, not argument. This shows that memo is optional

Comment (foxpy) [reply x/interchaintxs/client/cli/tx.go]: Well, if it is a common practice, then I am leaving an approving review. And at the same time anyone is free to change arguments hint if desirable.

**Review by foxpy — APPROVED** [review]

File: x/interchaintxs/client/cli/tx.go (hunk 3 above)
Comment (NeverHappened) [reply x/interchaintxs/client/cli/tx.go]: Changed to flag, did I do it ok?)

**Review by swelf19 — APPROVED** [review]

**Review by zavgorodnii — APPROVED** [review]
i have no time to test it like i did the last time, but i really hope that this works.

File: x/interchaintxs/client/cli/tx.go (hunk 3 above)
Comment (foxpy) [reply x/interchaintxs/client/cli/tx.go]: looks alright, though @pr0n00gler told us that:
> It's a common practice in Cosmos-SDK to wrap arguments in square brackets
so I am not 100% sure.

@pr0n00gler what do you think about it?

Comment (mmikev) [reply x/interchaintxs/client/cli/tx.go]: I think we should return square brackets here for consistency

File: x/interchaintxs/client/cli/tx.go [hunk 4]
```diff
@@ -67,7 +69,7 @@ func RegisterInterchainAccountCmd() *cobra.Command {
 
 func SubmitTxCmd() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "submit-tx [connection-id] [owner] [path/to/sdk_msgs.json]",
+		Use:     "submit-tx [connection-id] [owner] [path/to/sdk_msgs.json] --memo memo",
```
Comment (sotnikov-s) [inline x/interchaintxs/client/cli/tx.go]: this seems to be redundant cause you add the flag at the line 143

File: x/interchaintxs/client/cli/tx.go [hunk 5]
```diff
@@ -16,6 +16,8 @@ import (
 	"github.com/lidofinance/gaia-wasm-zone/x/interchaintxs/types"
 )
 
+const Memo = "memo"
```
Comment (sotnikov-s) [inline x/interchaintxs/client/cli/tx.go]: please add a Flag suffix to the constant name cuz otherwise it's misleading what is the `Memo` in the code

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Pull request merged**